### PR TITLE
GeoPointMapActivity crash when map isn't ready resolved

### LINKED ISSUE #2291: GeoPointMapActivity's Location Listener triggering a NullPointerException
#### Software and hardware versions 
Collect v1.15.0, Android v4.4.2

#### Problem description
`markerOptions` is null until the `setupMap` function of `GeoPointMapActivity` is called. The `LocationClients` is registering a listener before the `setupMap` function and if a location event is triggered then the `markerOptions` is being utilized before it's instantiation.

#### Steps to reproduce the problem
Get a location change event triggered before the map has loaded and that will cause a crash.

#### Expected behavior
` locationClient.setListener(this);` should be called after the `setupMap` function since all the location changes are dependent on the map being present so something like this should work.

```
 locationClient = LocationClients.clientForContext(this);
isMapReady = false;
((SupportMapFragment)getSupportFragmentManager().findFragmentById(R.id.map)).getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(GoogleMap googleMap) {
                setupMap(googleMap);
                locationClient.setListener(GeoPointMapActivity.this);
            }
        });
    }
```

#### Other information 

[Firebase bug link (authorized access)](https://console.firebase.google.com/u/0/project/api-project-322300403941/crashlytics/app/android:org.odk.collect.android/issues/5b204db06007d59fcda5bcc8?time=last-seven-days&sessionId=5B204C2602E200010AE40FEF6CBD0D0F_DNE_0_v2)

```
org.odk.collect.android.activities.GeoPointMapActivity.onLocationChanged (GeoPointMapActivity.java:396)
org.odk.collect.android.location.client.AndroidLocationClient.onLocationChanged (AndroidLocationClient.java:154)
android.location.LocationManager$ListenerTransport._handleMessage (LocationManager.java:279)
android.location.LocationManager$ListenerTransport.access$000 (LocationManager.java:208)
android.location.LocationManager$ListenerTransport$1.handleMessage (LocationManager.java:224)

```

### PR DESCRIPTION
**Pull request by jd-alexander**
Closes #2291

#### What has been done to verify that this works as intended?
The 	`GeoPointMapActivity` was opened and location events were received when the `onMapReady` functionality wasn't triggered and no crash occurred because the `locationClient` only registers a listener when the map is ready.

#### Why is this the best possible solution? Were any other approaches considered?
This is best solution because the code inside `onLocationChanged` relies on the map being available.

#### Are there any risks to merging this code? If so, what are they?
No.

#### Do we need any specific form for testing your changes? If so, please attach one.
The default widget forms can test this. Just navigate to the GeoPoint Widget question.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew pmd checkstyle lint findbugs` and confirmed all checks still pass.
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/opendatakit/collect/blob/master/CONTRIBUTING.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java [hunk 1]
```diff
@@ -219,89 +213,67 @@ public void onClick(View v) {
...
-                marker = null;
-                setClear = false;
-                latLng = new LatLng(location.getLatitude(), location.getLongitude());
-                markerOptions.position(latLng);
-                if (marker == null) {
-                    marker = map.addMarker(markerOptions);
-                    if (draggable && !readOnly) {
-                        marker.setDraggable(true);
-                    }
+        reloadLocation.setOnClickListener(v -> {
```
Comment (huangyz0918) [inline collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java:217]: I have a question, for Collect code, should we use lambda for all those methods? I found lots of lambda expressions but there are still many methods not be refactored with that. Do you think we have the necessary to keep the consistency? We can open a new issue maybe.

Comment (jd-alexander) [reply collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java:217]: I honestly think we should refactor as we touch code that have these listeners. Most of it is boilerplate so the lambda makes things more concise. Also the codebase will always have a mixture of lambda and normal expressions so it should be fine 😊

Comment (huangyz0918) [reply collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java:217]: 👍 Great, @jd-alexander . Can you help to do this refactor? You can see #2294

File: collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java [hunk 2]
```diff
@@ -496,7 +466,7 @@ public void onCancel(DialogInterface dialog) {
             if (location != null) {
                 zoomLocationButton.setEnabled(true);
                 zoomLocationButton.setBackgroundColor(Color.parseColor("#50cccccc"));
-                zoomLocationButton.setTextColor(themeUtils.getPrimaryTextColor());
+                zoomLocationButton.setTextColor(Color.parseColor("#ff333333"));
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java]: These changes aren't lambda changes so probably shouldn't be in this commit (and likely not this PR?).

File: collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java [hunk 3]
```diff
@@ -581,12 +547,11 @@ public String getLocationStatus() {
     }
 
     public String getAccuracyStringForLocation(Location location) {
-        return getString(R.string.location_provider_accuracy, GeoPointUtils.capitalizeGps(location.getProvider()),
```
Comment (lognaturel) [inline collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java]: This call was desired to make gps -> GPS. Is there a reason you've removed it? If you do change behavior, please make sure that there's an explanation somewhere (commit messages ideally or maybe PR description).

**Review by lognaturel — COMMENTED** [review]
I've marked a couple of seemingly-unrelated changes inline. This also doesn't compile due to missing resources. I haven't dug into it yet.

File: collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java (hunk 2 above)
Comment (jd-alexander) [reply collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java]: Thanks for spotting these.These changes aren't supposed to here so something got messed up when I was cloning master so making the fixes again.

File: collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java (hunk 3 above)
Comment (jd-alexander) [reply collect_app/src/main/java/org/odk/collect/android/activities/GeoPointMapActivity.java]: Same as above.

**Comment by jd-alexander** [comment]
@lognaturel thanks! I resolved the issues.

**Comment by lognaturel** [comment]
Awesome! And what was the PMD error? It could alternately be suppressed but I think your approach is a good one.

**Comment by jd-alexander** [comment]
Yes, it could be suppressed. I was using `GeoPointMapActivity.this` inside a callback and it didn't like that. It wanted me to use `this` only but that would reference the callback instead of the activity itself.

**Comment by lognaturel** [comment]
The issue is [UselessQualifiedThis](https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/unnecessary.html#UselessQualifiedThis). This is fixed in the latest version of PMD so instead of working around it, it might make sense to suppress it and then issue another PR updating the PMD version and fixing or excluding new checks that fail (`NcssCount` and `ClassNamingConventions` it looks like).

I updated the pmd version by adding
```
pmd {
    toolVersion = '6.4.0'
}
```
after `apply plugin: 'pmd'` in quality.gradle.

**Comment by jd-alexander** [comment]
> The issue is UselessQualifiedThis.

Yes that's the name of it. 

Okay will do the update.

**Comment by shobhitagarwal1612** [comment]
> Yes, it could be suppressed. I was using GeoPointMapActivity.this inside a callback and it didn't like that. It wanted me to use this only but that would reference the callback instead of the activity itself.

Yes, this is correct if it was an anonymous class but writing `this` in a lambda expression would reference the Activity and not the callback.

Quoted from https://docs.oracle.com/javase/tutorial/java/javaOO/lambdaexpressions.html

```
Accessing Local Variables of the Enclosing Scope
Like local and anonymous classes, lambda expressions can capture variables; they have the same access to local variables of the enclosing scope. However, unlike local and anonymous classes, lambda expressions do not have any shadowing issues (see Shadowing for more information). Lambda expressions are lexically scoped. This means that they do not inherit any names from a supertype or introduce a new level of scoping. Declarations in a lambda expression are interpreted just as they are in the enclosing environment. The following example, LambdaScopeTest, demonstrates this:
```



Hence writing this will definitely work as well.
```
((SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map)).getMapAsync(googleMap -> {	                      
     setupMap(googleMap);
     locationClient.setListener(this);
  });
```

So we should probably keep the check as it is correct.

**Comment by lognaturel** [comment]
@shobhitagarwal1612 the old version of PMD we are using flags false positives. If we update pmd we can then re-enable the check and it should pass. I have verified this with pmd 6.4.0.

**Comment by jd-alexander** [comment]
@lognaturel I started looking at the latest PMD after doing the update. I am seeing the `NcssCount` and `ClassNamingConventions` and tons of depreciation warnings so I am gonna go back to it soon when I wrap up some other things because a lot of rules have been impacted by the update.

**Comment by lognaturel** [comment]
👍 No rush!

**Comment by lognaturel** [comment]
The fix itself is going to be a little tough to verify because the crash only happened when a location change event happened to occur before full map load. @jd-alexander, I think a quick regression check on geopoint map is sufficient, does that sound ok?

**Comment by jd-alexander** [comment]
@lognaturel yes that sounds okay. What I did to test it was to trigger location events with `setupMap` commented out and that caused the crash so moving the listener stopped that, so I guess that's good enough.

**Comment by mmarciniak90** [comment]
Tested with success!

I was not able to reproduce the crash on the version without fix.
I did not notice regression. I verified all `Geo` widgets available in All widgets form

Verified on Android 4.1, 4.2, 4.4, 5.1, 6.0, 7.0, 8.1

@opendatakit-bot label "behavior verified"
@opendatakit-bot unlabel "needs testing"

**Review by lognaturel — APPROVED** [review]

**Pull request merged**